### PR TITLE
[ValueTracking] Simplify uaddo pattern

### LINKED ISSUE #65863: Simplify add-with-overflow pattern (uaddo)
For unsigned integers to check if an addition overflowed (get carry bit) it is enough to compare the sum with one of its arguments `carry = sum < a`. However, it is not obvious that it is enough to only check one of the arguments and we may encounter a pattern where both arguments are checked.

```llvm
define i1 @src(i64 %a, i64 %b){
  %s = add i64 %a, %b
  %cond_a = icmp ult i64 %s, %a
  %cond_b = icmp ult i64 %s, %b
  %cond = or i1 %cond_a, %cond_b
  ret i1 %cond
}

define i1 @tgt(i64 %a, i64 %b){
  %s = add i64 %a, %b
  %cond_a = icmp ult i64 %s, %a
  ret i1 %cond_a
}
```

Playground: https://godbolt.org/z/TW8enc6zo
Proof: https://alive2.llvm.org/ce/z/iPAE7t

### PR DESCRIPTION
**Pull request by dtcxzyw**
This patch simplifies the overflow check of unsigned addition.

`a + b <u a` implies `a + b <u b`
`a + b >=u a` implies `a + b >=u b`

Alive2: https://alive2.llvm.org/ce/z/H8oK8n
Fixes #65863.

### CODE REVIEW

**Comment by RKSimon** [comment]
Please can you pre-commit the tests to show the IR diff?

**Comment by dtcxzyw** [comment]
Ping.

File: llvm/lib/Analysis/ValueTracking.cpp [hunk 1]
```diff
@@ -8291,6 +8291,29 @@ static std::optional<bool> isImpliedCondICmps(const ICmpInst *LHS,
...
+    }
+    if (L1 == R1) {
+      std::swap(L0, L1);
+      LPred = ICmpInst::getSwappedPredicate(LPred);
+      std::swap(R0, R1);
+      RPred = ICmpInst::getSwappedPredicate(RPred);
+    }
+    if (L0 == R0 &&
+        (LPred == ICmpInst::ICMP_UGE || LPred == ICmpInst::ICMP_ULT) &&
+        (RPred == ICmpInst::ICMP_ULT || RPred == ICmpInst::ICMP_UGE) &&
```
Comment (goldsteinn) [inline llvm/lib/Analysis/ValueTracking.cpp:8312]: nit: `(RPred == ICmpInst::ICMP_UGE || RPred == ICmpInst::ICMP_ULT)` so the two lines match.

File: llvm/lib/Analysis/ValueTracking.cpp [hunk 2]
```diff
@@ -8291,6 +8291,29 @@ static std::optional<bool> isImpliedCondICmps(const ICmpInst *LHS,
...
+      LPred = ICmpInst::getSwappedPredicate(LPred);
+    }
+    if (L1 == R1) {
+      std::swap(L0, L1);
+      LPred = ICmpInst::getSwappedPredicate(LPred);
+      std::swap(R0, R1);
+      RPred = ICmpInst::getSwappedPredicate(RPred);
+    }
+    if (L0 == R0 &&
+        (LPred == ICmpInst::ICMP_UGE || LPred == ICmpInst::ICMP_ULT) &&
```
Comment (goldsteinn) [inline llvm/lib/Analysis/ValueTracking.cpp:8311]: what about `ugt` and `ule`?

Comment (dtcxzyw) [reply llvm/lib/Analysis/ValueTracking.cpp]: It is not true for `L0 >u L1 -> R0 >u R1` and  `L0 <=u L1 -> R0 <=u R1`.
Alive2: https://alive2.llvm.org/ce/z/VKtFkD

Comment (goldsteinn) [reply llvm/lib/Analysis/ValueTracking.cpp]: Works if you have `nsw` or `nuw`: https://alive2.llvm.org/ce/z/kzGqL4

Comment (dtcxzyw) [reply llvm/lib/Analysis/ValueTracking.cpp]: It is strange to detect unsigned addition overflows with `nsw` or `nuw` flag.
It doesn't work even if we have `nsw` or `nuw`: https://alive2.llvm.org/ce/z/eWsJH2.

**Comment by goldsteinn** [comment]
Okay, rebased ontop of your vec support patches + added vec support/tests.

**Review by goldsteinn — APPROVED** [review]

**Comment by goldsteinn** [comment]
LGTM.

**Pull request merged**